### DISCUSSION TITLE
Lifespan patch / mock in async testing

### DISCUSSION THREAD

**Discussion by tiptenbrink** [encode/starlette#2088] [Potential Issue] [answered]
Apologies if this is in a way a duplicate of https://github.com/encode/starlette/discussions/2084, but my problem is related more to accessing it / changing it in tests and how to activate it at all in async tests. Before, with startup/shutdown, the state was globally accessible from the app object. You could then modify the state in a test to refer to mock objects instead. I don't want to spin up a database in my tests so this is required.  Something like this:

```python
@pytest_asyncio.fixture(scope="module", autouse=True)
async def mock_dsrc(app_mod, app, api_config, module_mocker: MockerFixture):
    app.state.dsrc.gateway = module_mocker.MagicMock(spec=app.state.dsrc.gateway)
```

However, as far as I'm aware, you can't access the lifespan from the app object and since you are using the TestClient you cannot modify the individual requests either. I also don't know how to call the lifespan in the first place when using httpx's AsyncClient for async tests as recommended in the docs. 

I can think of some work-arounds, like wrapping every state access in a function that I can then replace by a mock, but I feel there should be a better way, or at least some documentation. I would love to be pointed in the right direction or help out with a solution.

Thanks for this great library :)

**Comment by adriangb** [comment, marked as answer]
Sorry for the delay in answering.

I generally try to structure things to avoid monkey patching. Take a look at this repo and see if that helps: https://github.com/adriangb/misc/tree/starlette-state-lifespan

I do think it would be low effort to add an argument to TestClient such as `with TestClient(app, lifespan_override=<a context manager>)` such that it'd be easy to override the lifespan with the `build_app()` pattern. `TestClient` would just not call the ASGI lifespan and instead call the custom lifespan and pass that state to requests.

**Reply by tiptenbrink** [reply]
Thanks a lot for the example! I was hoping to avoid having to make any changes to the app itself, but you're right this is probably the best way. Would it be useful to mention anything about this in the docs? (Maybe even the example you made) I remember when I originally tried to set up a testing architecture I had quite some difficulty and examples like this could help a lot.

**Reply by adriangb** [reply]
I think including docs for setting up a testable app architecture makes sense. @Kludex what do you think of converting the above example repo into something we include in the docs?

**Reply by Kludex** [reply]
I think we should include `lifespan_override(s)`.

Your example is great, but it also has the meaning of "you should now use this pattern", and that's opinionated - I don't think we should. I also don't see a cleaner way to do that without changing the way you create your application.

**Reply by tiptenbrink** [reply]
@adriangb I realized one of my questions above wasn't answered, namely how to test lifespan at all in async tests. The docs indicate to use httpx's AsyncClient directly, but this doesn't run the lifespan. Is there a way to manually run it and initialize it for an app?

Example:

```python
@pytest_asyncio.fixture(scope="module")
async def async_client():
    app = Starlette(lifespan=mock_lifespan)

    async with AsyncClient(app=app, base_url="http://test") as test_client:
        yield test_client
```

Doesn't run the lifespan and as a result making any request to an endpoint that depends on the lifespan will fail.

**Reply by adriangb** [reply]
https://www.python-httpx.org/async/#startupshutdown-of-asgi-apps

**Comment by havardthom** [comment]
Just to clarify, is there any way to access lifespan state from the app object ?

In my case I'm doing actual integration tests and would like to provide the apps database client through a session fixture, so I can forexample clear database between tests etc.
Previously I would just return the database client through `app.state.db`, but with the above approach I would have to duplicate the whole lifespan method in `conftest.py`.

_edit:_
I guess I could recreate lifespan state in a fixture, though I would prefer to reuse the state which is used in the app.
```
from starlette.datastructures import State

from src.lifespan import lifespan


@pytext.fixture(scope="session")
async def lifespan_state(app: FastAPI):
    async with lifespan(app) as state:
        yield State(state)
```

_edit2:_
For people using [asgi-lifespan](https://github.com/florimondmanca/asgi-lifespan) to run lifespan during tests, the state is available in the manager after v2.1.0 release
```
from typing import AsyncGenerator

import pytest
from asgi_lifespan import LifespanManager
from asgi_lifespan._types import ASGIApp
from fastapi import FastAPI
from httpx import AsyncClient
from starlette.datastructures import State

from src.app import create_app


@pytest.fixture(scope="session")
async def configured_app() -> FastAPI:
    app = create_app()
    return app

@pytest.fixture(scope="session")
async def lifespan_manager(configured_app: FastAPI) -> AsyncGenerator[LifespanManager, None]:
    async with LifespanManager(configured_app) as m:
        yield m

@pytest.fixture(scope="session")
async def lifespan_state(lifespan_manager: LifespanManager) -> State:
    return State(lifespan_manager._state)

@pytest.fixture(scope="session")
async def initialized_app(lifespan_manager: LifespanManager) -> ASGIApp:
    return lifespan_manager.app

@pytest.fixture(scope="session")
async def client(initialized_app: ASGIApp) -> AsyncGenerator[AsyncClient, None]:
    async with AsyncClient(app=initialized_app) as c:
        yield c
```
